File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if(s=="")
        return 0;
        int l =0,r=0;
        int n =s.length();
        int max_len = INT_MIN;
        unordered_map<char,int> mp;
        while(r<n)
        {
            if(mp.find(s[r])==mp.end())
            {
                max_len = max(max_len,r-l+1);
                mp[s[r]] = r;
                r++;
            }
            else{
                    l = max(mp[s[r]]+1,l);
                    mp[s[r]] = r;
                    max_len = max(max_len,r-l+1);
                    r++;
            }
        }
        return max_len;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        my_dict = {}

        l,r=0,0
        n=len(s)
        max_len = 0

        while(r<n):
         if s[r] in my_dict:

            l = max(my_dict[s[r]]+1,l)
            my_dict[s[r]]=r
            max_len = max(max_len,r-l+1)
            
         else:

            max_len = max(max_len,r-l+1)
            my_dict[s[r]]=r
         r+=1
        return max_len
        

